jobrunner/tracing.py

```diff
@@ -0,0 +1,226 @@
+
+    if "OTEL_EXPORTER_CONSOLE" in os.environ:
+        add_exporter(ConsoleSpanExporter())
+
+
+def trace_attributes(job):
+    """These attributes are added to every span in order to slice and dice by
+    each as needed.
+    """
+    # grab job request metadata, caching it on the job instance to avoid excess
+    # queries/jsoning
+    if job._job_request is None:
+        try:
+            job._job_request = database.find_one(
+                SavedJobRequest, id=job.job_request_id
+            ).original
+        except ValueError:
+            job._job_request = {}
+
+    attrs = dict(
+        backend=config.BACKEND,
+        job=job.id,
+        job_request=job.job_request_id,
+        workspace=job.workspace,
+        action=job.action,
+        commit=job.commit,
+        run_command=job.run_command,
+        user=job._job_request.get("created_by", "unknown"),
+        project=job._job_request.get("project", "unknown"),
+        orgs=",".join(job._job_request.get("orgs", [])),
+        state=job.state.name,
+        message=job.status_message,
+    )
+
+    if job.action_repo_url:
+        attrs["reusable_action"] = job.action_repo_url
+        if job.action_commit:
+            attrs["reusable_action"] += ":" + job.action_commit
+
+    # opentelemetry cannot handle None values
+    return {k: "" if v is None else v for k, v in attrs.items()}
+
+
+def initialise_trace(job):
+    """Initialise the trace for this job by creating a root span.
+
+    We store the serialised trace context in the db, so we can reuse it for
+    later spans.
+
+    We create a root span, which is a requirement in OTEl. For this reason we
+    send it out straight away, which means its duration is very short.
+    """
+    assert not job.trace_context, "this job already has a trace-context"
+    assert job.status_code is not None, "job has no initial StatusCode"
+    assert (
+        job.status_code_updated_at is not None
+    ), "job has no initial status_code_updated_at"
+
+    job.trace_context = {}
+    tracer = trace.get_tracer("jobs")
+    attrs = trace_attributes(job)
+
+    # convert created_at to ns
+    start_time = int(job.created_at * 1e9)
+
+    # TraceContextTextMapPropagator only works with the current span, sadlmy
+    with tracer.start_as_current_span("job", start_time=start_time) as root:
+        root.set_attributes(attrs)
+        # we serialise the entire trace context, as it may grow extra fields
+        # (e.g.  baggage) over time
+        TraceContextTextMapPropagator().inject(job.trace_context)
+
+    # trace the initial job state trace
+    start_new_state(job, job.status_code_updated_at)
+
+
+def finish_current_state(job, timestamp_ns, final=False, error=None, **attrs):
+    """Record a span representing the state we've just exited."""
+    try:
+        name = job.status_code.name
+        # Note: this *must* be timestamp as integer nanoseconds
+        start_time = job.status_code_updated_at
```

Comment:
wondering whether renaming this var to `job.status_code_updated_at_ns` could remove some of these comments?

Reply:
hmmm, some other things are nanoseconds but not `_ns~`, maybe just leave it as-is